aws/resource_aws_cloudhsm2_cluster_test.go

```diff
@@ -0,0 +1,100 @@
+package aws
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/hashicorp/terraform/helper/acctest"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func TestAccAWSCloudHsm2Cluster_basic(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSCloudHsm2ClusterDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccAWSCloudHsm2Cluster(),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSCloudHsm2ClusterExists("aws_cloudhsm_v2_cluster.cluster"),
+					resource.TestCheckResourceAttrSet("aws_cloudhsm_v2_cluster.cluster", "cluster_id"),
+					resource.TestCheckResourceAttrSet("aws_cloudhsm_v2_cluster.cluster", "vpc_id"),
+					resource.TestCheckResourceAttrSet("aws_cloudhsm_v2_cluster.cluster", "security_group_id"),
+					resource.TestCheckResourceAttrSet("aws_cloudhsm_v2_cluster.cluster", "cluster_state"),
+				),
+			},
+		},
+	})
+}
+
+func testAccAWSCloudHsm2Cluster() string {
+	return fmt.Sprintf(`
+variable "subnets" {
+  default = ["10.0.1.0/24", "10.0.2.0/24"]
+  type    = "list"
+}
+
+data "aws_availability_zones" "available" {}
+
+resource "aws_vpc" "cloudhsm2_test_vpc" {
+  cidr_block = "10.0.0.0/16"
+
+  tags {
+    Name = "terraform-testacc-aws_cloudhsm_v2_cluster-resource-basic"
+  }
+}
+
+resource "aws_subnet" "cloudhsm2_test_subnets" {
+  count                   = 2
+  vpc_id                  = "${aws_vpc.cloudhsm2_test_vpc.id}"
+  cidr_block              = "${element(var.subnets, count.index)}"
+  map_public_ip_on_launch = false
+  availability_zone       = "${element(data.aws_availability_zones.available.names, count.index)}"
+
+  tags {
+    Name = "tf-acc-aws_cloudhsm_v2_cluster-resource-basic"
+  }
+}
+
+resource "aws_cloudhsm_v2_cluster" "cluster" {
+  hsm_type = "hsm1.medium"  
+  subnet_ids = ["${aws_subnet.cloudhsm2_test_subnets.*.id}"]
+  tags {
+    Name = "tf-acc-aws_cloudhsm_v2_cluster-resource-basic-%d"
+  }
+}
+`, acctest.RandInt())
+}
+
+func testAccCheckAWSCloudHsm2ClusterDestroy(s *terraform.State) error {
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aws_cloudhsm_v2_cluster" {
+			continue
+		}
+		cluster, err := describeCloudHsm2Cluster(rs.Primary.ID, testAccProvider.Meta())
+
+		if err != nil {
+			return err
+		}
+
+		if cluster != nil && aws.StringValue(cluster.State) != "DELETED" {
+			return fmt.Errorf("CloudHSM cluster still exists:\n%s", cluster)
+		}
+	}
+
+	return nil
+}
+
+func testAccCheckAWSCloudHsm2ClusterExists(name string) resource.TestCheckFunc {
```

Comment:
This function should use `describeCloudHsm2Cluster()` to check the API that the cluster exists as expected.